File: recursion/linear_acc/applied/reverse_list.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
    def reverseList(head: ListNode): ListNode = {
        def recursiveReverseList(head : ListNode, acc : ListNode): ListNode = {
            head match {
                case null => acc
                case head => recursiveReverseList(head.next, ListNode(head.x, acc))
            }
        }
        recursiveReverseList(head, null)
    }
}

File: data_structures/two_pointers/standard/reverse_list.go
package data_structures

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// reverse linked list can either use an accumulator
// or can swap pointers
//
// x[1, y] -> y[2, z] -> z[3, nil]
func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	cur := head
	for cur != nil {
		next := cur.Next
		// swap
		cur.Next = prev

		// update
		prev = cur
		cur = next
	}
	return prev
}
